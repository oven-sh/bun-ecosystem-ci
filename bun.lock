{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "ecosystem-ci",
      "dependencies": {
        "@buildkite/buildkite-sdk": "^0.2.0",
        "commander": "^13.1.0",
      },
      "devDependencies": {
        "@types/bun": "latest",
        "oxlint": "^0.15.6",
        "prettier": "^3.4.2",
      },
      "peerDependencies": {
        "typescript": "^5.0.0",
      },
    },
  },
  "packages": {
    "@buildkite/buildkite-sdk": ["@buildkite/buildkite-sdk@0.2.0", "", { "dependencies": { "tslib": "^2.3.0", "yaml": "^2.6.1" } }, "sha512-vwyPd/HokZIzYT4283fFBoTPMmQPP6z82b92Yr/nl7rfh3iOgSPgQhDkdMdLiLEv7ZuSEbRKOyloMbZKromKyA=="],

    "@oxlint/darwin-arm64": ["@oxlint/darwin-arm64@0.15.6", "", { "os": "darwin", "cpu": "arm64" }, "sha512-Oyx21zpBMttAykoL3VSXM4ScfO0OygKAVFJlRFzh14CF4pnMQRcQ8fIH38BXy6wdE3YyZ9ml1ZspsCwguqb/3A=="],

    "@oxlint/darwin-x64": ["@oxlint/darwin-x64@0.15.6", "", { "os": "darwin", "cpu": "x64" }, "sha512-fNjxp7MBflKYyvvYgvjvAaCnVdYE3QZkscNhd8ygMko4yFErzPnvPong04QhvNN/cSxrL74nowMHrijPxZbm6Q=="],

    "@oxlint/linux-arm64-gnu": ["@oxlint/linux-arm64-gnu@0.15.6", "", { "os": "linux", "cpu": "arm64" }, "sha512-/yLShYUGVQV/lCxKDEKaZr5+QQm68NKuFkLcB3z0Qcd45BVdpa/0HGoOpOL7mexo+zxZkRQSGn7x2idVYv2Tng=="],

    "@oxlint/linux-arm64-musl": ["@oxlint/linux-arm64-musl@0.15.6", "", { "os": "linux", "cpu": "arm64" }, "sha512-S0i/xb/TEep2CjZ11D4BxifsaX2a1LyuxW1hFcREMebtXJjpZMFfCNR7Cnwk7tp6eWiXLo3DJ/6iPLkxV6iivg=="],

    "@oxlint/linux-x64-gnu": ["@oxlint/linux-x64-gnu@0.15.6", "", { "os": "linux", "cpu": "x64" }, "sha512-BNSRAn2SMF2/TVpQ/Y/Sg4emXia3Kfihu2apVFLyRj2h2lJCB0g/Us3gjRux2AoD4KkdlMJiqY/Tnaz2xMq49w=="],

    "@oxlint/linux-x64-musl": ["@oxlint/linux-x64-musl@0.15.6", "", { "os": "linux", "cpu": "x64" }, "sha512-BUv/dJIBXeinL2vgy1CvzwqbvnDfydU10urbofOF8XXob6AmDewZSIr3ctDvd5qJisR7oEhAuxH4+rPkKnmjEw=="],

    "@oxlint/win32-arm64": ["@oxlint/win32-arm64@0.15.6", "", { "os": "win32", "cpu": "arm64" }, "sha512-keVVbVxWBZ4myHlokg1JDQ0Fw2+iXMJSwbM3yr5SXXtMm7UgNaO3fGVAFA4Kv9OovhVCpfeqZy63gIgSv/w3oA=="],

    "@oxlint/win32-x64": ["@oxlint/win32-x64@0.15.6", "", { "os": "win32", "cpu": "x64" }, "sha512-onKba1QPlcX6sVBMp8KqsAmjgZfaanFZcYGYl6o6q6dzuZSLuKWCPyn/OeJ1PYryYVsbDKZYaCzLoBLrY13MqQ=="],

    "@types/bun": ["@types/bun@1.1.16", "", { "dependencies": { "bun-types": "1.1.43" } }, "sha512-E+ue6NMcn4FXC5bDRE1W/BXUVs01h5Mt02qH8/8HGCox9akuh8KNOFdwvaQS9TDgT2RmUyJYFRRqA60WtTnm2g=="],

    "@types/node": ["@types/node@20.12.14", "", { "dependencies": { "undici-types": "~5.26.4" } }, "sha512-scnD59RpYD91xngrQQLGkE+6UrHUPzeKZWhhjBSa3HSkwjbQc38+q3RoIVEwxQGRw3M+j5hpNAM+lgV3cVormg=="],

    "@types/ws": ["@types/ws@8.5.13", "", { "dependencies": { "@types/node": "*" } }, "sha512-osM/gWBTPKgHV8XkTunnegTRIsvF6owmf5w+JtAfOw472dptdm0dlGv4xCt6GwQRcC2XVOvvRE/0bAoQcL2QkA=="],

    "bun-types": ["bun-types@1.1.43", "", { "dependencies": { "@types/node": "~20.12.8", "@types/ws": "~8.5.10" } }, "sha512-W0wCtVH+bwFp7p3Zgs03CqxEDmXxEvmmUM/FBKgWIv9T8gyeotvIjIbHzuDScc2DphhRNtr7hJLCR5PspYL5qw=="],

    "commander": ["commander@13.1.0", "", {}, "sha512-/rFeCpNJQbhSZjGVwO9RFV3xPqbnERS8MmIQzCtD/zl6gpJuV/bMLuN92oG3F7d8oDEHHRrujSXNUr8fpjntKw=="],

    "oxlint": ["oxlint@0.15.6", "", { "optionalDependencies": { "@oxlint/darwin-arm64": "0.15.6", "@oxlint/darwin-x64": "0.15.6", "@oxlint/linux-arm64-gnu": "0.15.6", "@oxlint/linux-arm64-musl": "0.15.6", "@oxlint/linux-x64-gnu": "0.15.6", "@oxlint/linux-x64-musl": "0.15.6", "@oxlint/win32-arm64": "0.15.6", "@oxlint/win32-x64": "0.15.6" }, "bin": { "oxlint": "bin/oxlint", "oxc_language_server": "bin/oxc_language_server" } }, "sha512-LiDI0mIzvnL/oQq7AyAganWAKtk/RaiOKdACnzeJDe64gWp7c+0y3T554IvfWd/9kVt5awlvyg5eRRp67PaxNA=="],

    "prettier": ["prettier@3.4.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-e9MewbtFo+Fevyuxn/4rrcDAaq0IYxPGLvObpQjiZBMAzB9IGmzlnG9RZy3FFas+eBMu2vA0CszMeduow5dIuQ=="],

    "tslib": ["tslib@2.8.1", "", {}, "sha512-oJFu94HQb+KVduSUQL7wnpmqnfmLsOA/nAh6b6EH0wCEoK0/mPeXU6c3wKDV83MkOuHPRHtSXKKU99IBazS/2w=="],

    "typescript": ["typescript@5.7.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-84MVSjMEHP+FQRPy3pX9sTVV/INIex71s9TL2Gm5FG/WG1SqXeKyZ0k7/blY/4FdOzI12CBy1vGc4og/eus0fw=="],

    "undici-types": ["undici-types@5.26.5", "", {}, "sha512-JlCMO+ehdEIKqlFxk6IfVoAUVmgz7cU7zD/h9XZ0qzeosSHmUJVOzSQvvYSYWXkFXC+IfLKSIffhv0sVZup6pA=="],

    "yaml": ["yaml@2.7.1", "", { "bin": { "yaml": "bin.mjs" } }, "sha512-10ULxpnOCQXxJvBgxsn9ptjq6uviG/htZKk9veJGhlqn3w/DxQ631zFF+nlQXLwmImeS5amR2dl2U8sg6U9jsQ=="],
  }
}
