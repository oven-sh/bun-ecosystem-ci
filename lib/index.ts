export type { Context, EcosystemSuite } from './test-suite'
export { TestSuite, TestCase, Step } from './test-suite'
export * as steps from './steps'

// test suite templates
export { installAndTest } from './install-and-test'
