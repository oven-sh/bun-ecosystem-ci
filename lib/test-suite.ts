import deepmerge from 'deepmerge'
import type { Maybe } from './types'
import { strict as assert } from 'node:assert'
import { truthy } from '../src/util'

/**
 * Test suites have similar checks over a bunch of cases. Each case normally
 * corresponds to a single NPM package.
 *
 * These types are designed to be mappable to both `bun test` and
 * buildkite.
 * - {@link TestSuite} -> workflow
 * - {@link TestCase} -> job
 * - {@link Step} -> step lol
 */
export interface Context<D = any> {
    /**
     * Are we running tests locally or in CI?
     */
    isLocal: boolean
    runner: 'bun' | 'shell' | 'buildkite'
    /**
     * Name of bun binary. Use this instead of hardcoding `'bun'` into steps.
     */
    bun: string
    data?: D
}

export type EcosystemSuite<D = any> =
    | TestSuite<D>
    | ((context: Readonly<Context<D>>) => TestSuite<D> | Promise<TestSuite<D>>)

export interface TestSuite<D = any> {
    name?: string
    cases: TestCase[]
    beforeAll?: (context: Readonly<Context<D>>) => void | Promise<void>
    afterAll?: (context: Readonly<Context<D>>) => void | Promise<void>
}
export namespace TestSuite {
    export async function reify(
        suite: EcosystemSuite,
        context: Readonly<Context>
    ): Promise<TestSuite> {
        const testSuite =
            typeof suite === 'function' ? await suite(context) : suite
        return testSuite
    }
}

export interface TestCase {
    /**
     * Display name of the test case.
     */
    name: string
    steps: Step[]
    /**
     * Default working directory for all {@link Step steps}. Steps may
     * individually override this.
     *
     * @default process.cwd()
     */
    cwd?: string
    /**
     * This case is currently not passing.
     *
     * Test cases marked as failing that actually pass will fail (to force you
     * to update `failing` to `false`).
     * @default false
     */
    failing: boolean
    /**
     * Do not run this test case. Takes precedence over {@link failing}.
     * @default false
     */
    skip: boolean
    /**
     * Environment variables to set for each step.
     *
     * Overrides existing values in {@link process.env}
     */
    env?: Record<string, string | undefined>
}
export namespace TestCase {
    export interface Options extends Partial<Omit<TestCase, 'name' | 'steps'>> {
        /**
         * The steps taken by this test case. Steps are normalized with {@link Step.from}.
         */
        steps: Maybe<string | Step>[]
    }

    /**
     * Create a test case that runs a series of steps.
     *
     * @param name Name of the test case
     * @param steps Steps to run. Strings are treated as shell commands.
     * @returns A test case
     */
    export function from(name: string, steps: Maybe<string | Step>[]): TestCase
    /**
     * Create a test case configured with options.
     *
     * @param name Name of the test case
     * @param testCase Options for the test case
     * @returns A test case
     */
    export function from(name: string, testCase: Options): TestCase
    export function from(
        name: string,
        stepsOrOptions:
            | Options
            | Maybe<string | Step | Maybe<string | Step>[]>[]
    ): TestCase {
        const {
            steps,
            cwd,
            env,
            failing = false,
            skip = false,
        } = Array.isArray(stepsOrOptions)
            ? { steps: stepsOrOptions }
            : stepsOrOptions

        return {
            name,
            steps: steps
                .flat()
                .filter(truthy)
                .map(s => Step.from(s)),
            cwd,
            env,
            failing,
            skip,
        }
    }
}

export interface Step {
    /**
     * Unique identifier for the step. Should be in `kebab-case`.
     * @example "install-bun"
     */
    key?: string
    /**
     * Human-friendly display name
     */
    name?: string
    /**
     * Each line is a shell command to run
     */
    run: string[]
    /**
     * Environment variables to set for this step.
     *
     * Overrides existing values in {@link process.env} and {@link TestCase.env}
     */
    env?: Record<string, string | undefined>
    /**
     * Current working directory to use when running commands for this step.
     */
    cwd?: string
    /**
     * Timeout for this step in milliseconds. If the step takes longer than
     * this, it fails. By default, no timeout is set.
     *
     * @min 1
     */
    timeout?: number
    /**
     * Buildkite-specific configuration. These aren't used for other kinds of runners.
     */
    buildkite?: Step.BuildkiteOptions
}
export namespace Step {
    export type Options = Partial<Omit<Step, 'run'>>

    export interface BuildkiteOptions {
        /**
         * [Buildkite plugins](https://buildkite.com/docs/pipelines/integrations/plugins)
         * to add to the step.
         */
        plugins?: Record<string, any>
        artifactPaths?: string[]
    }

    export function from(
        command: string | string[] | Step,
        rest: Options = {}
    ): Step {
        if (Step.is(command)) {
            for (const key in rest) {
                const k = key as keyof Options
                const value = rest[k]
                if (value == null) continue
                const existing = command[k]
                if (
                    existing &&
                    typeof existing === 'object' &&
                    typeof value === 'object'
                ) {
                    command[k] = deepmerge(existing as any, value as any) as any
                } else {
                    command[k] ||= value as any
                }
            }
            return command
        }

        return {
            ...rest,
            run: Array.isArray(command)
                ? command
                : command
                      .trim()
                      .split('\n')
                      .map(s => s.trim())
                      .filter(truthy),
        }
    }

    /**
     * @returns `true` if `value` is a {@link Step}
     */
    export function is(value: unknown): value is Step {
        if (typeof value !== 'object' || !value) return false
        if (!('run' in value) || !Array.isArray(value.run)) return false
        if ('name' in value) {
            const name = value.name
            if (name !== undefined && typeof name !== 'string') return false
        }

        return true
    }

    export function validate(step: Step): void {
        assert(step.run?.length, 'Step must have at least one command')
        if (step.timeout != null) {
            assert(
                step.timeout > 0,
                `Timeout must be greater than 0, got ${step.timeout}`
            )
        }
    }
}
