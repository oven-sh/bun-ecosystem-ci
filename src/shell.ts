import assert from 'node:assert'
import type { Step, TestCase, TestSuite, Maybe } from '../lib'
import { truthy } from './util'

/**
 * Render a test suite into a shell script.
 *
 * @internal
 *
 * @param suite The test suite to render
 */
export function renderSuite(suite: TestSuite): string[] {
    const { name, cases } = suite
    assert(cases.length > 0, `Suite '${name}' has no test cases`)

    const renderedCases: string[] = cases
        .map(renderTestCase)
        .map(steps => (steps.push('\n'), steps))
        .flat()

    return [
        '#!/bin/bash',
        `# Suite: ${name}`,
        '',
        'set -eo pipefail',
        '',
        ...renderedCases,
    ]
}

/**
 * Render a test case into a shell script.
 *
 * @internal
 *
 * @param testCase the test case to render
 * @returns a list of lines. Join them together to form part of, or a whole,
 * bash script.
 */
export function renderTestCase(testCase: TestCase): string[] {
    const { name, cwd, steps, skip } = testCase
    assert(steps.length > 0, `Test case '${name}' has no steps`)

    let lines = [
        `# Test Case: ${name}`,
        `echo 'Running Test Case: ${name.replaceAll("'", "\\'")}'`,
        ...withDir(cwd)(steps.flatMap(renderStep)),
    ]

    if (skip) {
        lines = [`# Skipping '${name}'`, ...comment(lines)]
    }

    return lines
}

/**
 * Render a step as a list of shell commands.
 *
 * @internal
 *
 * @param step The step to render. Steps must have at least one command.
 * @returns a list of shell commands. These can be joined with newlines to
 * create part of a bash script.
 */
export function renderStep(step: Step): string[] {
    const { name, cwd, env, run } = step
    assert(run.length > 0, `Step '${name || '<anonymous>'}' has no commands`)

    let cmds: string[] = []
    if (cwd) cmds.push(`cd ${cwd}`)
    if (env)
        for (const [key, value] of Object.entries(env)) {
            cmds.push(`export ${key}="${value}"`)
        }
    cmds.push(...run)

    const lines = [`# Step: ${name ?? [run[0]]}`]
    if (name) lines.push(`echo '${name.replaceAll("'", "\\'")}'`)
    lines.push(...withTimeout(step.timeout)(subshell(indent(cmds))))

    return lines
}

type Lines = string[] | (() => string[])
const reifyLines = (lines: Lines) => (Array.isArray(lines) ? lines : lines())
const wrap = (enter: string, exit: string) => (lines: Lines) => [
    enter,
    ...reifyLines(lines),
    exit,
]

/**
 * Comment out commands
 */
const comment = (lines: Lines) => reifyLines(lines).map(line => `# ${line}`)
const indent = (lines: Lines) => reifyLines(lines).map(line => INDENT + line)
const INDENT = '  '

/** Change directories for the duration of some commands */
const withDir = (dir: string | undefined) =>
    dir ? wrap(`pushd ${dir}`, 'popd') : reifyLines //wrap('', '')
/** Runs commands in a subshell */
const subshell = wrap('(', ')')
const withTimeout =
    (timeout: Maybe<number>) =>
    (lines: string[]): string[] => {
        if (timeout == null || !lines.length) return lines
        if (lines.length > 2 && lines[0] == '(' && lines.at(-1) == ')') {
            let [_, ...rest] = lines
            rest.pop()
            if (!rest.length) return lines
            lines = rest
        }
        assert(timeout > 0)
        assert(lines.length > 0)
        const cmdArg = lines.filter(truthy).join(' && ').trim()
        return [
            `timeout --verbose --kill-after=${msToS(timeout * 2)} ${msToS(timeout)} bash -c "${cmdArg}"`,
        ]
    }
const msToS = (ms: number) => `${ms / 1000}s`
