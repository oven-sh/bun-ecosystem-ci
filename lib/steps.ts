import os from 'node:os'
import fs from 'fs'
import { Step, type Context } from './test-suite'

type Checkout = {
    repository: string
    ref?: string
    packageName: string
    isLocal: boolean
}
export const checkout: (options: Checkout) => Step = ({
    repository,
    ref = 'main',
    packageName,
    isLocal,
}): Step => {
    const clone = /* sh */ `git clone ${repository} --branch ${ref} --depth 1 ${packageName}`
    const script = !isLocal
        ? clone
        : /* sh */ `
if [ -d ${packageName} ]; then
    echo "resetting to ${ref}"
    cd ${packageName}
    git fetch --depth 1 --tags --prune-tags origin ${ref}
    git reset --hard ${ref}
else
    echo "cloning ${repository}/${ref}"
    ${clone}
fi
`.trim()

    return Step.from(script, {
        name: `Checkout ${packageName}`,
        key: `checkout-${packageName}`,
    })
}

export namespace test {
    /**
     * `bun test`
     */
    export const bun =
        (reportName = 'bun-test') =>
        ({ bun }: Context) => {
            // TODO: determine OS tmpdir
            const outfile = `/tmp/${reportName}.junit.xml`
            return Step.from(
                `${bun} test --reporter=junit --reporter-outfile=${outfile}`,
                {
                    name: 'bun test',
                    buildkite: {
                        artifactPaths: [outfile],
                        // plugins: {
                        //     'junit-annotate#v2.6.0': {
                        //         artifacts: 'bun-test.junit.xml',
                        //     },
                        // },
                    },
                }
            )
        }
}
