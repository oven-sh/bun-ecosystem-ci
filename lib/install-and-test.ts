import assert from 'node:assert'
import fs from 'fs'
import type { Context, EcosystemSuite } from './test-suite'
import { TestCase, Step } from './test-suite'
import * as steps from './steps'

interface Package {
    /**
     * URL to git repo. This gets cloned
     */
    repository: string
    /**
     * Ref to check out (branch name, commit, tag, etc).
     * @default "main"
     */
    ref?: string
    /**
     * How to run tests. "bun" is prepended implicitly.
     * @default "run test"
     */
    test?: string | string[]
    preload?: string | string[]
}

/**
 * Create an ecosystem test suite that 
 * 1. Checks out a repository
 * 2. Installs dependencies (`bun install`)
 * 3. Runs all of its tests (`bun run test`)
 * 
 * @param packages Packages to clone and test. Each one gets turned into a {@link TestCase}.
 */
export function installAndTest(
    packages: Record<string, Package>
): EcosystemSuite {
    return function doInstallAndTest({ bun }) {
        const defaultTestStep = `${bun} run test`

        const cases = Object.entries(packages).reduce(
            (
                cases: TestCase[],
                [packageName, { test, repository, ref = 'main', preload }]
            ) => {
                let testStep: string = defaultTestStep
                if (typeof test === 'string') {
                    assert(test.length, `test step cannot be empty`)
                    testStep = `${bun} ${test}`
                } else if (Array.isArray(test)) {
                    assert(test.length, `test step cannot be empty`)
                    assert.notEqual(test[0], 'bun')
                    const cleanTests = test.map(arg =>
                        arg.includes(' ') ? `"${arg}"` : arg
                    )
                    testStep = [bun, ...cleanTests].join(' ')
                }

                let bunfig = /* toml */ `[run]\nbun=true\n`
                if (preload) {
                    bunfig += '[test]\n'
                    if (Array.isArray(preload)) {
                        bunfig +=
                            'preload = [' +
                            preload.map(p => `\\"${p}\\"`).join(', ') +
                            ']\n'
                    } else {
                        bunfig += `preload = \\"${preload}\\"`
                    }
                }

                const testCase = TestCase.from(packageName, [
                    steps.checkout({ packageName, ref, repository }),
                    Step.from(`
                    if [ -d ${packageName} ]; then
                        cd ${packageName}
                        git reset --hard ${ref}
                    else
                        git clone ${repository} --branch ${ref} --depth 1 --no-tags ${packageName}
                    fi
                    `),
                    Step.from(`echo "${bunfig}" > bunfig.toml`, {
                        name: 'Create bunfig',
                        cwd: packageName,
                    }),
                    Step.from(`${bun} install`, { cwd: packageName }),
                    Step.from(testStep, { cwd: packageName }),
                ])

                cases.push(testCase)
                return cases
            },
            [] as TestCase[]
        )

        return {
            cases,
            beforeAll({ isLocal }: Context) {
                if (isLocal) {
                    try {
                        fs.mkdirSync('repos', { recursive: true })
                    } catch (e) {
                        const err = e as NodeJS.ErrnoException
                        if (err.code !== 'EEXIST') throw err
                    }
                    process.chdir('repos')
                }
            },
        }
    }
}
