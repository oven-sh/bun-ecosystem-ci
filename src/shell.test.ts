import { describe, beforeAll, it, expect } from 'bun:test'
import os from 'node:os'
import path from 'node:path'
import { renderStep, renderSuite, renderTestCase } from './shell'
import { Step, TestCase, type Context } from '../lib'
import foundationRegression from '../suites/foundation-regression'
import { AssertionError } from 'node:assert'

const id = () => Math.round(Math.random() * 10_000).toString(16)
const getTmpdir = (prefix: string) =>
    path.join(os.tmpdir(), 'ecosystem-ci', 'shell', prefix + '-' + id())

let shellcheck: string | undefined
beforeAll(async () => {
    try {
        shellcheck = await Bun.$`which shellcheck`.text()
        console.log(shellcheck)
    } catch {
        switch (process.platform) {
            case 'darwin':
                await Bun.$`brew install shellcheck`
                break
            case 'linux':
                await Bun.$`apt-get install shellcheck`.nothrow()
                break
            default:
                throw new Error(
                    `Could not find or install shellcheck binary. Please download it and try again.`
                )
        }
        shellcheck = await Bun.$`which shellcheck`.text()
    }
})

describe(renderSuite, () => {
    it('throws if no cases are provided', () => {
        expect(() => renderSuite({ cases: [] })).toThrow(AssertionError)
    })

    describe("When the 'foundation regression' suite is rendered", () => {
        let rendered: string[]
        let tmpdir: string
        const context: Readonly<Context> = Object.freeze({
            isLocal: true,
            bun: 'bun',
        })

        beforeAll(async () => {
            tmpdir = getTmpdir('render-suite')
            const suite =
                typeof foundationRegression === 'function'
                    ? await foundationRegression(context)
                    : foundationRegression
            rendered = renderSuite(suite)
        })

        it('produces a valid shell script', async () => {
            expect(shellcheck).toBeTruthy()
            const filename = path.join(tmpdir, 'basic-step.sh')
            await Bun.write(filename, rendered.join('\n'))
            const result = await Bun.spawn([shellcheck!, filename], {
                stdout: 'inherit',
            })
            expect(await result.exited).toBe(0)
        })
    })
})

describe(renderTestCase, () => {
    it('throws if no steps are provided', () => {
        expect(() => renderTestCase(TestCase.from('test', []))).toThrow(
            AssertionError
        )
    })
})

describe(renderStep, () => {
    it('throws if no commands are provided', () => {
        expect(() => renderStep({ run: [] })).toThrow(AssertionError)
    })

    describe('When a basic step is rendered', () => {
        let rendered: string[]
        let tmpdir: string
        const command = 'git clone git@github.com:oven-sh/bun-ecosystem-ci.git'

        beforeAll(() => {
            tmpdir = getTmpdir('render-step')
            rendered = renderStep(
                Step.from(command, { name: 'Clone this repository' })
            )
        })

        it('includes the step name', () => {
            expect(rendered[0]).toEqual('# Step: Clone this repository')
        })

        it('includes the step commands', () => {
            expect(rendered).toContainEqual(expect.stringMatching(command))
        })

        it('produces a valid shell script', async () => {
            expect(shellcheck).toBeTruthy()
            const filename = path.join(tmpdir, 'basic-step.sh')
            // steps are script fragments, not full scripts, so we need to add a
            // shebang. Otherwise shellcheck will complain
            await Bun.write(filename, '#!/bin/bash\n' + rendered.join('\n'))
            const result = await Bun.spawn([shellcheck!, filename], {
                stdout: 'inherit',
            })
            expect(await result.exited).toBe(0)
        })
    })
})
