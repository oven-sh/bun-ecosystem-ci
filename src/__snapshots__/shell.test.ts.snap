// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`renderStep respects \`step.timeout\` 1`] = `
"# Step: echo "hello world"
timeout --verbose --kill-after=2000 1000 (
  echo "hello world"
)"
`;
